feat: Changed the layout of logic analyzer graph fragment

harsh-2711 opened the pull request:
Fixes #889 #952 

Changes: Changed the layout of logic Analyzer fragment showing graph as decided in the issue tracker. Also, the layout was set to Landscape as the portrait mode doesn't suits the current layout

Screenshots for the change: 
![20180525_040343](https://user-images.githubusercontent.com/32356267/40518074-36e0d188-5fd6-11e8-9253-58197ee5bf39.gif)

APK for testing: 
[app-debug.zip](https://github.com/fossasia/pslab-android/files/2049172/app-debug.zip)

NOTE: Cursor Functionality will be added in subsequent PR

Vikum94 commented on app/src/main/java/org/fossasia/pslab/fragment/LALogicLinesFragment.java:
```diff
@@ -45,9 +49,13 @@
     private Runnable logicAnalysis;
+    private TextView channel1_text, channel2_text, channel3_text, channel4_text;
+    private LinearLayout Channel1, Channel2, Channel3, Channel4;
+    private String[] edges = {"EVERY EDGE", "EVERY FALLING EDGE", "EVERY RISING EDGE", "EVERY FOURTH RISING EDGE", "DISABLED"};
```
Attributes for the Spinner can be included in the resource file and populate it using that

harsh-2711 commented:
@Vikum94  @abhinavraj23  Changes done.

CloudyPadmal commented:
I'm a little confused about the edge select spinner list in the graph layout view. How `Logic Analyzer` works is as follows;

User can choose channels and select the edge they want. When he clicks on the "Analyze" button, PSLab will start capturing 2500 data points from ID pins. Once the collection is complete, those data points will be plotted on the graph plane. There after its not real-time.

Can you tell me the reason for adding those same selections in the main window also in the graph view as well?

harsh-2711 commented:
@CloudyPadmal  I didn't knew that the graph plotted wasn't real time for logic analyzer as it is for oscilloscope and so I thought to provide user the functionality to switch between selection of edges as it is provided in many other software in real time. So, keeping this layout aside, now only cursor and plotting of real pulses can be added as a feature. So, should I proceed with that?

CloudyPadmal commented:
@harsh-2711 sample size and frequency cannot be changed. It will fetch maximum samples at the maximum frequency. 

The reason behind logic analyzer being not a real-time instrument is that, in digital circuits we don't get periodical pulses often. We use a logic analyzer with multiple channels to test if the digital circuit output is responding correctly to it's many inputs at different stages. If this becomes real-time, there are too many data points which makes the analysis a lot harder. I think that explanation clears your doubts.

May be you can implement this missing feature in that space;
![timemeasurement](https://user-images.githubusercontent.com/14261304/40571766-5ea69e8c-60bc-11e8-8144-acb5695cb93b.png)

harsh-2711 commented:
@CloudyPadmal  Now I completely setup my environment on desktop and so now I can figure out what should be the layout as I tested the logic analyzer a lot.

harsh-2711 commented:
This is my final layout for Logic Analyzer
![20180529_042104](https://user-images.githubusercontent.com/32356267/40631522-37cbd458-62fb-11e8-99ee-05dff8d8800b.gif)

I have restricted the number of measurement cards to 20 and the output of every measurement will be visible on the right hand side of the measure button as soon as the measure button is pressed

CloudyPadmal commented:
@harsh-2711 they can be squashed. I taught @abhinavraj23 yesterday and he will be able to help you.

harsh-2711 commented:
@CloudyPadmal  @abhinavraj23 @Vikum94 @Avjeet  Changes done. Please review them.

abhinavraj23 commented on app/src/main/res/layout-land/logic_analyzer_logic_lines.xml:
```diff
@@ -1,42 +1,103 @@
+                    android:layout_marginTop="@dimen/chart_layout_margin_top"
+                    android:background="#000" />
+
+
```
Please remove these extra lines of code. Give it a code reformat :+1:

abhinavraj23 commented on app/src/main/res/layout-land/logic_analyzer_logic_lines.xml:
```diff
@@ -1,42 +1,103 @@
+
+        </LinearLayout>
+
+
```
Also here!

abhinavraj23 commented on app/src/main/res/layout/la_time_measurement_card.xml:
```diff
@@ -0,0 +1,91 @@
+        android:gravity="center_horizontal"
+        android:popupBackground="@color/colorPrimary" />
+
+
```
Why are we leaving two lines? Leaving one line is fine, so give all the files code reformat :+1:

harsh-2711 commented:
@abhinavraj23  I reformatted the code for all layouts. Please review them now.

CloudyPadmal commented:
@harsh-2711 why are we using a similar kind of a panel (the channel and edge select panel) in two windows? Can't we use only one? Also don't put that much border radius. Please go through Android Material Design guidelines.

harsh-2711 commented:
@CloudyPadmal  The panel used here is for time interval measurement and so contains different options than previous fragment. I will work on the border radius

CloudyPadmal commented:
@harsh-2711 we're not building the layout similar to desktop app. Remember we did a lot of research trying to find devices in market spending so many days. So lets make something make sense! :+1:

harsh-2711 commented:
@CloudyPadmal  I can't understand what is wrong with this layout as we need at least two edges to measure time interval between them and so I provided two Spinners for each section - one to choose channel ID and other to choose if it's rising,falling or fourth rising edge.

CloudyPadmal commented:
@harsh-2711 in your implementation, when we open Logic Analyzer, we get a screen to select channel modes and a button to start analyze.(this screen was there from the beginning) And when we click on analyze button, the graph window opens and it has the same set of mode select features. Why do we need the same thing in two places? Does that make sense to you?

All you've done in this PR is just adding a list of spinners which are already there in the instrument next to the graph area. I don't see any resemblance to the instrument image you've proposed in the beginning of this PR to current implementation. It looks more like the desktop app implementation.

harsh-2711 commented:
@CloudyPadmal I think you don't get my proposed PR. Here on pressing Measure button will measure the distance between two selected edges and will show next to measure button. Currently no text is seen as the code for calculating time difference isn't added. And multiple cards are given so that many measurements can be taken at a time between channel with any IDs. These spinners aren't same as the spinners provided in the instrument's main fragment. And I proposed the same layout before but only difference was that it had a start/stop button. You can further mention any ideas to be added and I will do that.

CloudyPadmal commented:
@harsh-2711 I'm not talking about a cursor measurement. I don't see any code doing any functionality. This is just a UI related PR right? So what I'm asking is why do we have to select the same stuff in two places? Why can't we have it in one place?

| First screen | Second screen |
| ---- | ---- |
| ![1](https://user-images.githubusercontent.com/14261304/40818315-3b495d96-6574-11e8-82d9-e2326a31ad2d.png) | ![2](https://user-images.githubusercontent.com/14261304/40818348-5bba76be-6574-11e8-9faf-3c82251e65bb.png) |

You have to think and do. It's really hard to review like this

harsh-2711 commented:
@CloudyPadmal  Ok I got the confusion. I was going to add the cursor functionality but in a different PR. But now I will add it in the current PR. Theses cards were meant to save the time of calculation to be done manually between edges. So my idea is I will keep the cards to measure the time difference between edges and cursor will be used to measure the time at any particular point. Is this implementation correct?

Vikum94 commented on app/src/main/res/layout/modified_spinner_dropdown_list.xml:
```diff
@@ -0,0 +1,12 @@
+    android:id="@android:id/text1"
+    style="?android:attr/spinnerDropDownItemStyle"
+    android:layout_width="match_parent"
+    android:layout_height="40dp"
```
This one can go in the resource file.

Vikum94 commented on app/src/main/res/values/dimens.xml:
```diff
@@ -109,6 +109,28 @@
+    <dimen name="chart_layout_margin_top">5dp</dimen>
+    <dimen name="chart_layout_margin_bottom">40dp</dimen>
+    <dimen name="recycler_view_margin_end">5dp</dimen>
+    <dimen name="button_text">13sp</dimen>
```
button_text_size would be a good name

Vikum94 commented:
Can there be more than 4 channel modes. When I keep on clicking New Measurement, there are new measure cards added. You have to limit that. 
Cannot go back to the first layout on back pressed, and as @CloudyPadmal mentioned, I agree on eliminating the first layout.

harsh-2711 commented:
@Vikum94  I have limited the cards to 20 as there are four channel modes and three possible edges type to measurement for each one. Also the back button wasn't working due to wrong commit. I fixed that and @CloudyPadmal here is the new layout in which I added cursor
![20180602_025355](https://user-images.githubusercontent.com/32356267/40864288-d49716cc-6610-11e8-9282-3fc5fe1bb962.gif)
As the video quality was poor so highlighter isn't visible to a greater extent. But it is visible on mobile. Here is the apk to test 
[app-debug.zip](https://github.com/fossasia/pslab-android/files/2064209/app-debug.zip)

I will soon make commits to make the cards functional so that we can decide whether to keep them or not.

CloudyPadmal commented:
@harsh-2711 please remove the first layout and have user directly come to the graph view. This is taking so long.

Please follow up on these issues;

1. Have you gone through PSLab-desktop-apps source code to see how Logic Analyzer works?
2. In the following figure do you see a cursor and have you checked the python source code in desktop app to see how we capture time intervals there?
![la](https://user-images.githubusercontent.com/14261304/40869182-da79b716-6633-11e8-8ac5-74289950e4b9.png)

FYI; we capture time intervals using the firmware and device itself. We don't have to explicitly calculate like how you are suggesting with a cursor.
```C++
case INTERVAL_MEASUREMENTS:
    lsb = getInt();
    location = getChar();
    value = getChar();
    Interval(location & 0xF, (location >> 4)&0xF, value & 0x7, (value >> 3)&0x7);
    while ((!_IC1IF) && (IC2TMR < lsb))asm("CLRWDT");
    sendInt(IC1BUF);
    sendInt(IC2BUF);
    while ((!_IC3IF) && (IC2TMR < lsb))asm("CLRWDT");
    sendInt(IC3BUF);
    sendInt(IC4BUF);
    sendInt(IC2TMR);
    disable_input_capture();
    break;
```

CloudyPadmal commented:
@harsh-2711 please focus on this instrument and complete this UI part soon. We are already at the end of [Milestone - 1](https://github.com/fossasia/pslab-android/milestone/17)

harsh-2711 commented:
@CloudyPadmal I know that how are we measuring time intervals in pslab-desktop-apps but for current scenario the logic analyzer in pslab app isn't working. So how should I check if the code is working or not. There are direct methods available in ScienceLab class of r2r and f2f measurement but for that Logic Analyzer needs to be working.

CloudyPadmal commented:
@harsh-2711 yes Logic Analyzer is not functional and your task is to make it functional. Not just UI. For this you can go through the python code; both [pslab-desktop-apps](https://github.com/fossasia/pslab-desktop-apps/) and [pslab-python](https://github.com/fossasia/pslab-python/) to see how it's implemented in the desktop and implement that code in Java

CloudyPadmal commented:
@harsh-2711 make this PR a UI related one. Please remove the first screen (white one) and navigate user directly to graph screen just like oscilloscope. Then we can merge this PR and work on.

harsh-2711 commented:
@CloudyPadmal  Yes I know that I have to make Logic Analyzer functional as discussed before and I am daily working on it. But what I meant to say that is it can't be made functional before Milestone-1. So for now I will remove the first layout

CloudyPadmal commented:
Yes that's what I'm asking. Remove the first screen and complete the layout. That's the task for Milestone 1. As the next step will be functionalities.

Avjeet commented:
@harsh-2711 The floating cards on the right side doesn't look good can you try to make it like fixed panel like physical logical analyzer with all the buttons, knobs and control? And not just on the right side, you could use the space at the bottom.

harsh-2711 commented:
@Avjeet  I can't use the bottom panel as it will mix all the legends and cursor position indicator with the channel selection. Also to fit all the channels at once on the right panel, I have to use scroll bar and so fixed layouts won't look good. Also I only have to use two spinners for each channel and no knobs or control or buttons needs to be added.

CloudyPadmal commented on .travis.yml:
```diff
@@ -10,7 +10,7 @@ android:
     - android-27
     - build-tools-27.0.3
-   
+
```
Don't make any changes in `travis`

CloudyPadmal commented:
Can you add some margin between the tile and Analyze button. Also for channel selector what if use numbers instead of text? Those fonts look so compact.
![screenshot_2018-06-06-06-41-16](https://user-images.githubusercontent.com/14261304/41010460-3fef3994-6955-11e8-8b8f-7a796dbab1f1.png)

harsh-2711 commented:
@CloudyPadmal Changes done. Please review. My screen was big and so there was no problem of text in carousel view so I couldn't figure out. Thanks.

CloudyPadmal commented:
@harsh-2711 great! :+1: can you upload the APK so we can test and merge this to move forward? Also there are some conflicts as a PR got merged. Resolve that and build and APK and upload it here! :+1:

harsh-2711 commented:
Here is the apk to test [app-debug.zip](https://github.com/fossasia/pslab-android/files/2075800/app-debug.zip)

harsh-2711 commented:
@CloudyPadmal  I am not able to figure out why Travis is failing. Can you please help?

CloudyPadmal commented:
How do you conclude it's #960 ? I just ran #989 and the Travis build passed. I see other PRs also getting passed. Please double check your code Google the issue in stacktrace to figure out where the issue is.

CloudyPadmal commented on gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -3,4 +3,4 @@ distributionBase=GRADLE_USER_HOME
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-4.5.1-all.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-4.5.1-all.zip
```
Here you've made modifications to configurations. That's why I guess it's failing. As a developer try to figure these things out.

harsh-2711 commented:
@CloudyPadmal  I changed the gradle wrapper file to original but then also it fails. I searched for it a lot but there are same questions available and not answer on Gradle Forum. Please help. And I concluded that it was PR #960  because after I merged it in my branch and solved merge conflicts then this error started to appear.

harsh-2711 commented:
@CloudyPadmal I made a different branch and copied the same code and it worked. I am not able to figure out what's the problem with this branch. So, can I close this PR and open new one or should I continue searching for the bug? The code is exactly same and so doesn't needs review.

CloudyPadmal commented:
@harsh-2711 okay then open a new PR and reference this from there. Let's get this merged quickly as this is been there for long!

CloudyPadmal commented:
#1007 is opened with Travis build fix for this.
